tex/generic/pgf/utilities/pgfkeys.code.tex

```diff
@@ -901,7 +901,13 @@
   \pgfkeys@searchalso@parse#1,\pgfkeys@mainstop
   {%
     \toks0=\expandafter{\pgfkeys@global@temp##1\pgfeov}%
-  \toks1={\pgfkeysalso{/handlers/.unknown/.@cmd/.expand once=\pgfkeys@searchalso@temp@value}}%
+  \toks1={%
+    \pgfkeysgetvalue{/handlers/.unknown/.@cmd}{\pgfkeys@code}%
+    % trick:
+    % to allow unbalanced \if values (stored in \pgfkeys@searchalso@temp@value),
+    % expand the \fi right after "\the\toks1 " first, see below
+    \expandafter\expandafter\expandafter\pgfkeys@code\expandafter\pgfkeys@searchalso@temp@value\expandafter\pgfeov
+  }%
```

Comment:
Maybe something like this?  That avoids the non-local `\expandafter` chain.
```patch
--- a/tex/generic/pgf/utilities/pgfkeys.code.tex
+++ b/tex/generic/pgf/utilities/pgfkeys.code.tex
@@ -24,6 +24,8 @@
 % works with all TeX formats and has no save-stack impact
 
 \def\pgfkeys@empty{}
+\long\def\pgfkeys@firstoftwo#1#2{#1}
+\long\def\pgfkeys@secondoftwo#1#2{#2}
 
 % This is useful:
 
@@ -901,10 +903,14 @@
   \pgfkeys@searchalso@parse#1,\pgfkeys@mainstop
   {%
     \toks0=\expandafter{\pgfkeys@global@temp##1\pgfeov}%
-  \toks1={\pgfkeysalso{/handlers/.unknown/.@cmd/.expand once=\pgfkeys@searchalso@temp@value}}%
+  \toks1={\pgfkeysalso{/handlers/.unknown/.expand once=\pgfkeys@searchalso@temp@value}}%
   \xdef\pgfkeys@global@temp{%
     \noexpand\def\noexpand\pgfkeys@searchalso@temp@value{####1}%
     \noexpand\ifpgfkeysaddeddefaultpath
+      \noexpand\expandafter\noexpand\pgfkeys@firstoftwo
+    \noexpand\else
+      \noexpand\expandafter\noexpand\pgfkeys@secondoftwo
+    \noexpand\fi{%
       \noexpand\pgfkeyssuccessfalse
       \noexpand\let\noexpand\pgfkeys@searchalso@name=\noexpand\pgfkeyscurrentkeyRAW
       \the\toks0 % one or more /.try things; one for each path. The last element won't have a /.try
@@ -912,9 +918,9 @@
       %\noexpand\else
       %  \the\toks1 % invoke /handlers/.unknown handler
       %\noexpand\fi
-    \noexpand\else
+    }{%
       \the\toks1 % invoke /handlers/.unknown handler
-    \noexpand\fi
+    }%
   }%
   \expandafter\gdef\expandafter\pgfkeys@global@temp\expandafter##\expandafter1\expandafter\pgfeov\expandafter{\pgfkeys@global@temp}%
   }%
```

Reply:
That's clearer.

As discussed in https://github.com/pgf-tikz/pgf/issues/671#issuecomment-683338545, point 2, `\pgfkeysalso` should be avoided since it updates macros like `\pgfkeyscurrentkeysRAW`, which might be used in `/handlers/.unknown/.@cmd` to show info of unknown key.

I will apply your suggested changes except for
```diff
-  \toks1={\pgfkeysalso{/handlers/.unknown/.@cmd/.expand once=\pgfkeys@searchalso@temp@value}}%
+  \toks1={\pgfkeysalso{/handlers/.unknown/.expand once=\pgfkeys@searchalso@temp@value}}%
```
being replaced with
```diff
-  \toks1={\pgfkeysalso{/handlers/.unknown/.@cmd/.expand once=\pgfkeys@searchalso@temp@value}}%
+  \toks1={%
+    \pgfkeysgetvalue{/handlers/.unknown/.@cmd}{\pgfkeys@code}%
+    \expandafter\pgfkeys@code\pgfkeys@searchalso@temp@value\pgfeov
+  }%
```

Reply:
Sorry I didn't notice that the suggested change is already committed in https://github.com/pgf-tikz/pgf/commit/c78fc2ed513422b814f44d352f9691fb1e98d04a#diff-9e7dc0e51a095224bbb67963bebc4fb8.

But the above discussion about avoiding `\pgfkeysalso` in `\toks1={\pgfkeysalso{...}}` is still valid.